### PR TITLE
[BEAM-6431] Implement Execution Time metrics start,process,finish in the Java SDK

### PR DESCRIPTION
**Pull request by ajamato**
[BEAM-6431] Implement Execution Time metrics start,process,finish in the Java SDK

**Please** add a meaningful description for your change here

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | ---

### CODE REVIEW

**Comment by ajamato** [comment]
@ryan-williams @Ardagan 
https://github.com/apache/beam/pull/7676
This PR is ready for review now. Uses the ExecutionStateTracker in the Java SDK to collect PTransformExecutionTimeRegistry.

I also started using matches for testing MonitoringInfos, we can add more matchers over time, this eliminates the need to cleanup/clear timestamps, values, etc when testing monitoring infos.

The usage of the states and state tracker mostly happen inside the PCollectionConsumerRegistry and PTransformFunctionRegistry. Once again I have changed its signature and ctors, which affects many tests which are using them (due to the repeated patterns in tests)

PTAL and let me know what you think of these changes.

File: runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/SimpleExecutionState.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+ */
+package org.apache.beam.runners.core.metrics;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.beam.runners.core.metrics.ExecutionStateTracker.ExecutionState;
+
+/**
+ * Simple state class which collects the totalMillis spent in the state. Allows attaching an
```
Comment (Ardagan) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/SimpleExecutionState.java]: Would be good to specify what it attaches key value labels to.

File: runners/core-java/src/test/java/org/apache/beam/runners/core/metrics/MonitoringInfoMatchers.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+   *
+   * <p>This is useful for tests which do not want to match the specific value (execution times).
+   * Currently this will only check for URNs, labels, type URNs and int64Values.
+   */
+  public static TypeSafeMatcher<MonitoringInfo> matchSetFields(final MonitoringInfo mi) {
+    return new TypeSafeMatcher<MonitoringInfo>() {
+
+      @Override
+      protected boolean matchesSafely(MonitoringInfo item) {
+        // Match the URN.
```
Comment (Ardagan) [inline runners/core-java/src/test/java/org/apache/beam/runners/core/metrics/MonitoringInfoMatchers.java]: you can remove comments here.

File: runners/java-fn-execution/src/test/java/org/apache/beam/runners/fnexecution/control/RemoteExecutionTest.java [hunk 3]
```diff
@@ -510,30 +514,33 @@ public void testMetrics() throws Exception {
                 ParDo.of(
                     new DoFn<byte[], String>() {
                       private boolean emitted = false;
-                      Counter startCounter =
+                      private Counter startCounter =
                           Metrics.counter(RemoteExecutionTest.class, startUserCounterName);
 
                       @StartBundle
                       public void startBundle() throws InterruptedException {
+                        Thread.sleep(1000);
```
Comment (Ardagan) [inline runners/java-fn-execution/src/test/java/org/apache/beam/runners/fnexecution/control/RemoteExecutionTest.java]: any way for us to use smaller values here, or avoid Thread.sleep at all?

File: sdks/java/harness/src/main/java/org/apache/beam/fn/harness/data/PCollectionConsumerRegistry.java [hunk 4]
```diff
@@ -113,6 +134,12 @@ public PCollectionConsumerRegistry(MetricsContainerStepMap metricsContainerRegis
     return wrappedConsumer;
   }
 
+  /** @return Execution Time MonitoringInfos based on the tracked start or finish function. */
```
Comment (Ardagan) [inline sdks/java/harness/src/main/java/org/apache/beam/fn/harness/data/PCollectionConsumerRegistry.java]: duplicate comment

File: runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/SimpleExecutionState.java (hunk 1 above)
Comment (ajamato) [reply runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/SimpleExecutionState.java]: Done

File: runners/core-java/src/test/java/org/apache/beam/runners/core/metrics/MonitoringInfoMatchers.java (hunk 2 above)
Comment (ajamato) [reply runners/core-java/src/test/java/org/apache/beam/runners/core/metrics/MonitoringInfoMatchers.java]: Done

File: sdks/java/harness/src/main/java/org/apache/beam/fn/harness/data/PCollectionConsumerRegistry.java (hunk 4 above)
Comment (ajamato) [reply sdks/java/harness/src/main/java/org/apache/beam/fn/harness/data/PCollectionConsumerRegistry.java]: Done

File: runners/java-fn-execution/src/test/java/org/apache/beam/runners/fnexecution/control/RemoteExecutionTest.java (hunk 3 above)
Comment (ajamato) [reply runners/java-fn-execution/src/test/java/org/apache/beam/runners/fnexecution/control/RemoteExecutionTest.java]: The one way we could remove these sleeps is by exposing the metrics during the test, and polling until we see them. However, the metrics are only available when the bundle finishes. 

I filed BEAM-6597 with a TODO on how to do this and upgrade this test.

I don't want to go much lower than this, for fear of making this test noisy/flakey.

**Pull request merged**